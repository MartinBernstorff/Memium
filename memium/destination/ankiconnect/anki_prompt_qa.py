import html
from collections.abc import Sequence
from dataclasses import dataclass

import genanki

from memium.utils.extract_terms import get_terms_surrounded_by_underscores

from ...utils.hash_cleaned_str import hash_str_to_int
from .anki_prompt import AnkiPrompt


@dataclass(frozen=True)
class AnkiQA(AnkiPrompt):
    base_deck: str
    tags: Sequence[str]
    question: str
    answer: str
    css: str
    uuid: int  # UUID used for scheduling. If a new note is added with the same uuid, it will override the old note.

    @property
    def genanki_model(self) -> genanki.Model:
        model_fields = [
            {"name": "Question"},
            {"name": "Answer"},
            {"name": "Extra"},
            {"name": "UUID"},
        ]

        QUESTION_STR = r"{{ Question }}"
        TTS_QUESTION_STR = r"{{ tts en_US voices=Apple_Samantha speed=1.05:Question }}"

        ANSWER_STR = r"{{ Answer }}"
        TTS_ANSWER_STR = r"{{ tts en_US voices=Apple_Samantha speed=1.05:Answer }}"

        EXTRA_STR = r"{{ Extra }}"

        model_template = [
            {
                "name": "Ankdown QA Card with UUID",
                "qfmt": f"""
<div class="extra">
    {EXTRA_STR}
</div>
<div class="front">
    {QUESTION_STR}{TTS_QUESTION_STR}
</div>
            """,
                "afmt": f"""
<div class="back">
    <div class="extra">
        {EXTRA_STR}
    </div>
    <div class="question">
        {QUESTION_STR}
    </div>
    <div class="answer">
        {ANSWER_STR}{TTS_ANSWER_STR}
    </div>
</div>
            """,
            }
        ]

        return genanki.Model(
            model_id=hash_str_to_int("Ankdown QA with UUID"),
            name=("Ankdown QA with UUID"),
            fields=model_fields,
            templates=model_template,
            css=self.css,
            model_type=0,
        )

    @property
    def _extra_field_content(self) -> str:
        return "" if not self.edit_url else edit_button(self.edit_url)

    def to_genanki_note(self) -> genanki.Note:
        return genanki.Note(
            guid=str(self.uuid),
            model=self.genanki_model,
            fields=[
                html.escape(field)
                for field in [
                    self._field_to_markdown(self.question),
                    self._field_to_markdown(self.answer),
                    self._extra_field_content,
                    str(self.uuid),
                ]
            ],
            tags=self.tags,
        )

    @property
    def deck(self) -> str:
        base_deck = super().deck
        wiki_links = get_terms_surrounded_by_underscores(self.question)
        wiki_subdeck = "-".join(sorted(wiki_links))

        if wiki_subdeck != "":
            return f"{base_deck}::{wiki_subdeck}"
        return base_deck


def edit_button(url: str) -> str:
    return f"""<a href="{url}" style="background-color: #5f0069;
border: none;
color: white;
padding: 0.8em;
text-align: center;
text-decoration: none;
font-size: 0.8em;
font-family: 'Inter', sans-serif;
float: right;
border-radius: 8px;
opacity: 0.8;
">Obsidian</a>"""
