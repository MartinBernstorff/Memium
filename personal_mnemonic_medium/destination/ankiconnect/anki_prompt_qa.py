from collections.abc import Sequence
from dataclasses import dataclass

import genanki

from ...utils.hash_cleaned_str import clean_str, int_hash_str
from .anki_card import AnkiCard


@dataclass(frozen=True)
class AnkiQA(AnkiCard):
    base_deck: str
    tags: Sequence[str]
    question: str
    answer: str
    css: str

    @property
    def uuid(self) -> int:
        return int_hash_str(clean_str(f"{self.question}{self.answer}"))

    @property
    def genanki_model(self) -> genanki.Model:
        model_fields = [
            {"name": "Question"},
            {"name": "Answer"},
            {"name": "Extra"},
            {"name": "UUID"},
        ]

        QUESTION_STR = r"{{ Question }}"
        TTS_QUESTION_STR = (
            r"{{ tts en_US voices=Apple_Samantha speed=1.05:Question }}"
        )

        ANSWER_STR = r"{{ Answer }}"
        TTS_ANSWER_STR = (
            r"{{ tts en_US voices=Apple_Samantha speed=1.05:Answer }}"
        )

        EXTRA_STR = r"{{ Extra }}"

        model_template = [
            {
                "name": "Ankdown QA Card with UUID",
                "qfmt": f"""
<div class="front">
    {QUESTION_STR}{TTS_QUESTION_STR}
</div>
<div class="extra">
    {EXTRA_STR}
</div>
            """,
                "afmt": f"""
<div class="back">
    <div class="question">
        {QUESTION_STR}
    </div>
    <div class="answer">
        {ANSWER_STR}{TTS_ANSWER_STR}
    </div>
    <div class="extra">
        {EXTRA_STR}
    </div>
</div>
            """,
            }
        ]

        return genanki.Model(
            model_id=int_hash_str("Ankdown QA with UUID"),
            name=("Ankdown QA with UUID"),
            fields=model_fields,
            templates=model_template,
            css=self.css,
            model_type=0,
        )

    def to_genanki_note(self) -> genanki.Note:
        return genanki.Note(
            guid=str(self.uuid),
            model=self.genanki_model,
            fields=[
                self.question,
                self.answer,
                " ".join(self.tags),
                str(self.uuid),
            ],
            tags=self.tags,
        )
