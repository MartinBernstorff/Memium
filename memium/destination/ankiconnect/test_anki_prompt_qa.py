from collections.abc import Sequence
from dataclasses import dataclass, field

from .anki_prompt_qa import AnkiQA


def fake_tag_factory() -> Sequence[str]:
    return ["FakeTag"]


@dataclass(frozen=True)
class FakeAnkiQA(AnkiQA):
    base_deck: str = "FakeBaseDeck"
    tags: Sequence[str] = field(default_factory=fake_tag_factory)
    question: str = "FakeQuestion"
    answer: str = "FakeAnswer"
    css: str = "FakeCSS"


def test_ankiqa_deck_inference():
    card = FakeAnkiQA(tags=["anki/deck/Subdeck"])
    assert card.deck == "FakeBaseDeck::Subdeck"
