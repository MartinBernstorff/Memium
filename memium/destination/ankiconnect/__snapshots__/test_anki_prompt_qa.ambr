# serializer version: 1
# name: test_formatting
  list([
    '<p>Q. This is a <em>question</em>?</p>',
    '<p>FakeAnswer</p>',
    '''
      &lt;a href=&quot;FakeEditURL&quot; style=&quot;background-color: #5f0069;
      border: none;
      color: white;
      padding: 0.8em;
      text-align: center;
      text-decoration: none;
      font-size: 0.8em;
      font-family: &#x27;Inter&#x27;, sans-serif;
      float: right;
      border-radius: 8px;
      opacity: 0.8;
      &quot;&gt;Obsidian&lt;/a&gt;
    ''',
    '0',
  ])
# ---
